### PR TITLE
Feat: REPLAT-1403 - Article Summary Refactor

### PR DESCRIPTION
**Pull request by jimhunty**
As a result of Article Summary being needed in multiple places, this is being moved more towards a factory approach. This has been implemented and updated across all packages *other than slice*. This is waiting on the first PR into this feature branch.

### CODE REVIEW

File: packages/article-summary/__tests__/__snapshots__/article-summary.native.test.js.snap [hunk 1]
```diff
@@ -9,13 +9,18 @@ exports[`renders an article-summary component with a single paragraph 1`] = `
...
-        "fontFamily": "GillSansMTStd-Medium",
-        "fontSize": 12,
-        "letterSpacing": 1,
-        "marginBottom": 2,
-      }
+      Array [
+        Object {
+          "fontFamily": "GillSansMTStd-Medium",
+          "fontSize": 12,
+          "fontWeight": "400",
```
Comment (colincclark) [inline packages/article-summary/__tests__/__snapshots__/article-summary.native.test.js.snap]: how come the style changed?

File: packages/article-summary/__tests__/__snapshots__/article-summary.native.test.js.snap [hunk 2]
```diff
@@ -553,6 +573,20 @@ exports[`renders an article-summary component with headline and no content 1`] =
   >
     Saturday July 01 2017, 03:32pm, The Sunday Times
   </Text>
+  <Text
```
Comment (colincclark) [inline packages/article-summary/__tests__/__snapshots__/article-summary.native.test.js.snap]: do not agree with showing an empty Text component, should not render at all if emtpy

File: packages/article-summary/__tests__/__snapshots__/article-summary.native.test.js.snap [hunk 3]
```diff
@@ -746,6 +780,20 @@ exports[`renders an article-summary component with no byline 1`] = `
   >
     Friday November 17 2017, 12:01am, The Times
   </Text>
+  <Text
```
Comment (colincclark) [inline packages/article-summary/__tests__/__snapshots__/article-summary.native.test.js.snap]: Empty Text elements are not great. Prefer the whole element to not render when content or data is empty or null

File: packages/article-summary/__tests__/article-summary-helper.js [hunk 4]
```diff
@@ -1,12 +1,12 @@
 import React from "react";
 import renderer from "react-test-renderer";
-import defaultFixture from "../fixtures/default.json";
-import articleMultiFixture from "../fixtures/article-multi.json";
-import emptyParagraphFixture from "../fixtures/article-empty-paragraph.json";
-import noBylineFixture from "../fixtures/no-byline.json";
-import noLabelFixture from "../fixtures/no-label.json";
-import reviewFixture from "../fixtures/review.json";
-import blankFixture from "../fixtures/blank.json";
+import defaultFixture from "../fixtures/default";
```
Comment (colincclark) [inline packages/article-summary/__tests__/article-summary-helper.js]: i took from our discussion that we did _not_ want to make dynamic tests and therefore would leave these as `json` files? @craigbilner

File: packages/article-summary/article-summary.js [hunk 5]
```diff
@@ -35,85 +31,44 @@ const styles = {
...
-    showPublication,
-    byline
-  } = props;
-  const summary = summarise(text);
-  const labelText = label && label.toUpperCase && label.toUpperCase();
+  const { Label, Headline, textAst, DatePublication, Byline } = props;
 
   return (
     <View style={styles.container}>
-      {label ? <Text style={styles.label}>{labelText}</Text> : null}
```
Comment (colincclark) [inline packages/article-summary/article-summary.js]: you now have unused styles above

File: packages/article-summary/article-summary.js [hunk 6]
```diff
@@ -35,85 +31,44 @@ const styles = {
...
-      {headline ? (
-        <ArticleSummaryHeadline
-          hasResponsiveHeadline={hasResponsiveHeadline}
-          headline={headline}
-        />
-      ) : null}
-      <Text style={styles.text}>{renderTrees(summary, renderer)}</Text>
+      <Label />
+      <Headline />
+      <Text style={styles.text}>{textAst()}</Text>
```
Comment (colincclark) [inline packages/article-summary/article-summary.js]: Can you get the functions all working without the wrapping `Text`'s? Perhaps the Styled Components that are passed in can do the styling that you have added here

File: packages/article-summary/article-summary.js [hunk 7]
```diff
@@ -35,85 +31,44 @@ const styles = {
...
-  headline: PropTypes.string,
-  hasResponsiveHeadline: PropTypes.bool,
-  text: PropTypes.arrayOf(treePropType),
-  date: DatePublication.propTypes.date,
-  publication: DatePublication.propTypes.publication,
-  showPublication: DatePublication.propTypes.showPublication,
-  byline: PropTypes.arrayOf(treePropType)
+  Label: PropTypes.func,
+  Headline: PropTypes.func,
+  textAst: PropTypes.func,
```
Comment (colincclark) [inline packages/article-summary/article-summary.js]: can this be named something more meaningful? `summaryText` or some such thing

File: packages/article-summary/article-summary.js [hunk 8]
```diff
@@ -35,85 +31,44 @@ const styles = {
...
-  showPublication: DatePublication.defaultProps.showPublication,
-  byline: []
+  Label: () => null,
+  Headline: () => null,
+  textAst: () => [],
+  DatePublication: () => null,
+  Byline: () => null
 };
 
+export { summarise, renderer, ArticleSummaryHeadline };
```
Comment (colincclark) [inline packages/article-summary/article-summary.js]: I kind of get why these are exported from `article-summary`, but why don't the consumers import them directly rather than from here? This dynamic is a little confusing/misleading

File: packages/article-summary/fixtures/article-empty-paragraph.js [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+export default {
+  Label: () => <ArticleLabel title="Camilla Long" color="#333333" />,
+  DatePublication: () => (
+    <DatePublication
+      date="2017-07-01T14:32:00.000Z"
+      publication="SUNDAYTIMES"
+    />
+  ),
+  Headline: () => (
+    <ArticleSummaryHeadline headline="OK, so Putin’s not a lady, but he does have the wildest man‑PMT" />
```
Comment (colincclark) [inline packages/article-summary/fixtures/article-empty-paragraph.js]: lol @ Putin's PMT

File: packages/article-summary/__tests__/article-summary-helper.js (hunk 4 above)
Comment (colincclark) [reply packages/article-summary/__tests__/article-summary-helper.js]: I mean I understand why you added the functions to the fixtures, but maybe the stories should import the `json` fixture data, and handle the creation of functions as a normal consumer component would do?

File: packages/article-summary/summarise.js [hunk 10]
```diff
@@ -0,0 +1,14 @@
+export default function summarise(text) {
```
Comment (colincclark) [inline packages/article-summary/summarise.js]: this could benefit from a test, or are the snapshots sufficient?

File: packages/author-profile/author-profile-item.web.js [hunk 11]
```diff
@@ -61,21 +71,29 @@ const AuthorProfileItem = item => {
...
+      label ? <ArticleLabel title={label} color="#333333" /> : null,
+    Headline: () => <ArticleSummaryHeadline headline={headline} />,
+    DatePublication: () => (
+      <DatePublication date={publishedTime} publication={publicationName} />
+    )
   };
 
   const children = showImage ? (
-    <ArticleSummary {...childProps} text={summary} />
+    <ArticleSummary {...childProps} textAst={() => renderAst(summary)} />
```
Comment (colincclark) [inline packages/author-profile/author-profile-item.web.js]: this file does not seem to have PropTypes

File: packages/author-profile/author-profile-item.web.js [hunk 12]
```diff
@@ -5,7 +5,17 @@ import Card from "@times-components/card";
 import Link from "@times-components/link";
 import { withTrackEvents } from "@times-components/tracking";
 import withResponsiveStyles from "@times-components/responsive-styles";
-import ArticleSummary from "@times-components/article-summary";
+
+import ArticleLabel from "@times-components/article-label";
+import DatePublication from "@times-components/date-publication";
+import { renderTrees } from "@times-components/markup";
+import ArticleSummary, {
+  ArticleSummaryHeadline,
```
Comment (colincclark) [inline packages/author-profile/author-profile-item.web.js]: `ArticleLabel` is imported separately, so can we import these other components in the same way?

File: packages/author-profile/author-profile-item.web.js [hunk 13]
```diff
@@ -61,21 +71,29 @@ const AuthorProfileItem = item => {
...
   const childProps = {
-    label,
-    headline,
-    date: publishedTime,
-    publication: publicationName
+    Label: () =>
+      label ? <ArticleLabel title={label} color="#333333" /> : null,
+    Headline: () => <ArticleSummaryHeadline headline={headline} />,
+    DatePublication: () => (
+      <DatePublication date={publishedTime} publication={publicationName} />
```
Comment (colincclark) [inline packages/author-profile/author-profile-item.web.js]: So these endering functions should contain any positioning / containing elements rather than the empty ones being displayed when data is unavailable

File: packages/author-profile/author-profile-item.web.js [hunk 14]
```diff
@@ -61,21 +71,29 @@ const AuthorProfileItem = item => {
   }
 
   const childProps = {
-    label,
-    headline,
-    date: publishedTime,
-    publication: publicationName
+    Label: () =>
+      label ? <ArticleLabel title={label} color="#333333" /> : null,
```
Comment (colincclark) [inline packages/author-profile/author-profile-item.web.js]: lol, we seemed to have just moved the ternaries somewhere else 😉

**Comment by colincclark** [comment]
The other thing I would say about this PR is that I do not believe it should be folded into the feature branch. This is a separate refactor that does not deliver two related articles. It can be folded into master on it's own.

File: packages/article-summary/__tests__/__snapshots__/article-summary.native.test.js.snap (hunk 1 above)
Comment (jimhunty) [reply packages/article-summary/__tests__/__snapshots__/article-summary.native.test.js.snap]: This is because it uses the `<ArticleLabel />` component. Made more sense.

File: packages/article-summary/__tests__/__snapshots__/article-summary.native.test.js.snap (hunk 2 above)
Comment (jimhunty) [reply packages/article-summary/__tests__/__snapshots__/article-summary.native.test.js.snap]: The alternative is having a turnery in the code, which we were trying to avoid. The extra text is on Native and therefore wouldnt have any real impact.

File: packages/article-summary/__tests__/__snapshots__/article-summary.native.test.js.snap (hunk 3 above)
Comment (jimhunty) [reply packages/article-summary/__tests__/__snapshots__/article-summary.native.test.js.snap]: As above.

File: packages/article-summary/__tests__/article-summary-helper.js (hunk 4 above)
Comment (jimhunty) [reply packages/article-summary/__tests__/article-summary-helper.js]: No that wasn't the whole story. The fixtures needed to match the tests, that was what was agreed. These are not dynamic, the take in the same components every time. They will only change if the packages they derive from change. What you are suggesting will keep what we are doing but add an extra file.

File: packages/article-summary/article-summary.js (hunk 6 above)
Comment (jimhunty) [reply packages/article-summary/article-summary.js]: As we discussed, one of the responsibilities of this file would be to enforce some styling. If you were to pass through components with their own styling then this would overide.

File: packages/author-profile/author-profile-item.web.js (hunk 11 above)
Comment (jimhunty) [reply packages/author-profile/author-profile-item.web.js]: Nope, and it didnt have them before. A requirement for this Article Summary to pass or tackle this in another ticket?

File: packages/author-profile/author-profile-item.web.js (hunk 13 above)
Comment (jimhunty) [reply packages/author-profile/author-profile-item.web.js]: Are you referring to the empty Text from Byline? This is covered in another comment

File: packages/author-profile/author-profile-item.web.js (hunk 14 above)
Comment (jimhunty) [reply packages/author-profile/author-profile-item.web.js]: Not all Author Profile Items have a label. Can you suggest something else?

File: packages/author-profile/author-profile-item.web.js (hunk 12 above)
Comment (jimhunty) [reply packages/author-profile/author-profile-item.web.js]: Its a separate package.

File: packages/article-summary/article-summary.js (hunk 8 above)
Comment (jimhunty) [reply packages/article-summary/article-summary.js]: Makes for a cleaner import statement. Its a clearer API of what is available as its the top level file.

File: packages/author-profile/author-profile-item.web.js (hunk 14 above)
Comment (nikhedonia) [reply packages/author-profile/author-profile-item.web.js]: Alternative would be to make articleLabel#title optional

File: packages/article-summary/article-summary.js (hunk 7 above)
Comment (jimhunty) [reply packages/article-summary/article-summary.js]: Done

File: packages/article-summary/article-summary.js (hunk 5 above)
Comment (jimhunty) [reply packages/article-summary/article-summary.js]: Done

File: packages/article-summary/summarise.js (hunk 10 above)
Comment (jimhunty) [reply packages/article-summary/summarise.js]: Can we look at this once #588 is merged? Then we can check the coverage. I would like a test, but its a case of where to fit this in.

File: packages/author-profile/author-profile-item.web.js (hunk 14 above)
Comment (jimhunty) [reply packages/author-profile/author-profile-item.web.js]: Then we would get a blank article label component. I think we shouldnt change the implementation of that.

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/51d8ed2bd59690a0e39d82788c6f40d60e815077/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/725c6b4ad13562b0a8f7c7f8b9214091209a8562/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/1a92da389e2d9d7563d8439de33e6ba4dc6a316e/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/5c6a796e4fe323a3bcdb26d44c733a5c7da47c9b/dextrose-test/index.html

File: packages/author-profile/author-profile-item.web.js (hunk 14 above)
Comment (craigbilner) [reply packages/author-profile/author-profile-item.web.js]: I would say label and byline should handle their own empty states?

File: packages/author-profile/author-profile-item.web.js (hunk 13 above)
Comment (craigbilner) [reply packages/author-profile/author-profile-item.web.js]: as above would push this to the component level, consumers shouldn't care about this?

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/ce2f4cb8cbb94428ed2c45d703e05cf5580e7191/dextrose-test/index.html

**Comment by jimhunty** [comment]
@craigbilner This originally was and was in review for at least a day. We needed to make sure the testing on article summary was up to scratch so we moved it in. I will create another PR with that change.

File: packages/author-profile/author-profile-item.web.js (hunk 14 above)
Comment (aronshamash) [reply packages/author-profile/author-profile-item.web.js]: would be cleaner if `label` and `byline` handled I agree.

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/36e79e9ae12e1a50b95a6fb0e8977e2128d0bfe0/dextrose-test/index.html

File: packages/author-profile/author-profile-item.web.js (hunk 14 above)
Comment (colincclark) [reply packages/author-profile/author-profile-item.web.js]: `label` and `byline` can return `null` as appropriate. The consumer, i.e. this file, will then pass `null` down to `article-summary` which in turn will not render _anything_. The styled containing `View` and `Text` elements you have added (around Label and DatePublication) would then only get rendered if the relevant `prop` was not `null`.

Comment (colincclark) [reply packages/author-profile/author-profile-item.web.js]: Also, those styled containing elements arguably exist in the consumer, so article-summary is more re-usable

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/4832edea1e6bbda3d3b399d6eaa7c7e5b2e19c38/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/15afab3945823b6b34f0ab5c2fbda2d0566f06bb/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/802e2512c4bac483df99920f7a1593a365f8886f/dextrose-test/index.html

File: packages/article-summary/fixtures/blank.js [hunk 15]
```diff
@@ -0,0 +1,17 @@
+import React from "react";
```
[the remark this answers is not in the document]
Comment (aronshamash) [reply packages/article-summary/fixtures/blank.js]: This looks like an existing file name.

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/b49d7aa6bebe3b41368baff88cbc4d7a3f7f44b5/dextrose-test/index.html

**Comment by colincclark** [comment]
Also, not sure what is causing it, but the Author Profile stories seem to be broken and the Card component (in Card/Default story) has text now bleeding off of the top of the page....? See screenshots. 
screen shot 2018-02-08 at 10 43 12
screen shot 2018-02-08 at 10 43 22

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/abd52f6e39cc4fd48221a3809ae8fdf8eb78b023/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/5934eb54d627235b99db90ad885dae9eb6293f80/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/e1d335a2c02d4738ed7d93671960a2700385954e/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/01162d5005d8012c595f226002fe9668c7d67107/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/9fc1b4cbe88b8663d1a30636321932e81aafe55b/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/31a2cb752a1c110346fe7a5d9f11595cdb4ff4ca/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/e29e860481be0097d8270892a880e06a2417c344/dextrose-test/index.html

File: packages/article-summary/styles/index.web.js [hunk 16]
```diff
@@ -0,0 +1,3 @@
+import sharedStyles from "./shared";
```
Comment (colincclark) [inline packages/article-summary/styles/index.web.js]: no need for this file as it does the same as `index.js`

File: packages/article-summary/summarise.js (hunk 10 above)
Comment (colincclark) [reply packages/article-summary/summarise.js]: where? was expecting `summarise.test.js` or similar

File: packages/gradient/package.json [hunk 17]
```diff
@@ -4,10 +4,7 @@
   "description": "Renders a gradient background",
   "main": "gradient",
   "scripts": {
-    "test:android": "jest --config='./__tests__/android/jest.config.js'",
```
Comment (colincclark) [inline packages/gradient/package.json]: why were these removed...seems like they are still required? gradient has `__tests__/<platform>/...` directories

File: packages/slice/slice-content.web.js [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+    imageSize: 996,
+    showImage: true
+  };
+
+  return (
+    <Link url={url} onPress={onPress}>
+      <View style={styles.cardContainer}>
+        <Card {...cardProps} image={imageUri ? { uri: imageUri } : null}>
+          <ArticleSummary
+            bylineProps={{ ast: byline }}
```
Comment (colincclark) [inline packages/slice/slice-content.web.js]: I might make a fork of this today. I expected something totally different and do not feel like that is getting across. Byline should do its own "should I show" logic, returning null, when it should not show. The consumer, Slice in this case, should style the byline and pass it across (or null if that is what was returned from byline). ArticleSummary should just call a function which returns whatever was passed in. That is, a (styled) component or null.

Comment (jimhunty) [reply packages/slice/slice-content.web.js]: Speak to @craigbilner before you make too much effort as you were not involved in the latest discussion we had.

You either have a Byline or not. There will be no other changes in style to that Byline as its consistent. If at some point it requires a styled component then that is a new feature and would then represent Headline more. This goes the same with Label and DatePublication.

You would always want to show a Headline so there should be no "should i show" logic.

Comment (colincclark) [reply packages/slice/slice-content.web.js]: OK. Fair enough. This is a return to having ternaries in the article-summary component. I am not sure I see the value in the PR at this point - it is a big departure from the JIRA ticket's goal.

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/91ade4d431d0858856f9905c3ee8b47023a52ef6/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/a1a26db6d35cdd973a6fcdffd31eb6f67b54b56c/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/820b2be12c0ab731857236cc5fb733f30307d4ad/dextrose-test/index.html

File: packages/slice/slice-content.web.js (hunk 18 above)
Comment (craigbilner) [reply packages/slice/slice-content.web.js]: @colincclark this removes all ternaries from a consumption point of view? There's always going to be somewhere which needs to present differently based on the data?

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/a4031c8ea811957d632b073d563a40a9153979f8/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/081c06951ddfbf60ec8a513c029fe4c6b2e5fef8/dextrose-test/index.html
